### DISCUSSION TITLE
@remix-run/dev/server-build

### DISCUSSION THREAD

**Discussion by andridge** [remix-run/remix#6060] [Q&A]
Hi I'm trying to deploy a hydrogen oxygen shopify e-commerce site build using remix and Ive been stuck with this error for a while now :undefined ERROR	Error: @remix-run/dev/server-build is not meant to be used directly from node_modules. It exists to provide type definitions for a virtual module provided by the Remix compiler at build time.

my Import in my server.js file is :
```js
 // Virtual entry point for the app
import * as remixBuild from '@remix-run/dev/server-build';
```
my vercel.json file is:
```json
{
  "version": 2,
  "routes": [
    { "src": "/api/(.*)", "dest": "/api/$1" },
    { "src": "/", "dest": "/server.js" },
    { "src": "/(.*)", "dest": "/server.js" }
  ],
  "builds": [
    { "src": "server.js", "use": "@vercel/node" },
    {
      "src": "app/**",
      "use": "@vercel/static",
      "config": { "basePath": "dist/client" }
    },
    {
      "src": "./server",
      "use": "npm:@vercel/node",
      "config": { "basePath": "dist/worker", "entry": "./server" }
    },
    {
      "src": "./build.css",
      "use": "npm:build:css",
      "config": { "entry": "./build.css" }
    }
  ]
}
```
HELP PLEASE: 🙏 😭

**Comment by rossipedia** [comment]
TL;DR: That error message you're getting happens when you try to import `@remix-run/dev/server-build` without using the `server` option.

---

Are you using Remix to compile/bundle your server code? By default, Remix builds the server bundle as a module that has a few exports in it (the routes, the manifest, that kind of thing). This is passed off to the `createRemixHandler()` function that is exposed by the various platform packages (`@remix-run/express`, etc...). This is done typically by using `require()` or `import()` and passing the location of the server build, which by default is the `build` folder.

When you use Remix to bundle your server (by providing a `server` option in `remix.config.js`), then Remix bundles the server build _along with_ your server code into a single module/bundle. In that instance, you don't `require()` or `import()` from a path, since there's no longer any separation.

Instead, you import/require from `@remix-run/dev/server-build` and it resolves to the application code that remix builds (ie: the `app` folder).

**Reply by andridge** [reply]
<img width="1408" alt="Screenshot 2023-05-22 at 17 09 32" src="https://github.com/remix-run/remix/assets/46260701/3de32148-8eee-4f70-9d59-398333475aff">
Yes i'm using remix to bundle my server.Which works well locally and in development.I forgot to mention that its hosting it in Vercel.
<img width="1408" alt="Screenshot 2023-05-22 at 17 10 51" src="https://github.com/remix-run/remix/assets/46260701/4572c15a-db86-4a3d-8681-e1bfef230bcd">
<img width="1408" alt="Screenshot 2023-05-22 at 16 59 28" src="https://github.com/remix-run/remix/assets/46260701/2d2f6ec1-fc98-432a-8899-f09848f9a3c0"> together with this error for the function
undefined	ERROR	Error: @remix-run/dev/server-build is not meant to be used directly from node_modules. It exists to provide type definitions for a virtual module provided by the Remix compiler at build time.
    at Object.<anonymous> (/var/task/node_modules/@remix-run/dev/server-build.js:18:7)
    at Module._compile (node:internal/modules/cjs/loader:1254:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1308:10)
    at Module.load (node:internal/modules/cjs/loader:1117:32)
    at Function.Module._load (node:internal/modules/cjs/loader:958:12)
    at Module.require (node:internal/modules/cjs/loader:1141:19)
    at require (node:internal/modules/cjs/helpers:110:18)
    at Object.<anonymous> (/var/task/server.js:2:1)
    at Module._compile (node:internal/modules/cjs/loader:1254:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1308:10)
RequestId: 1783ea83-b83b-4479-8b41-30411e319977 Error: Runtime exited with error: exit status 1
Runtime.ExitError
I still don't know what might be the problem as im importing  @remix-run/dev/server-build virtually and not from the modules directly.
My server file

import * as remixBuild from '@remix-run/dev/server-build';
import {createStorefrontClient, storefrontRedirect} from '@shopify/hydrogen';
import {
  createRequestHandler,
  getStorefrontHeaders,
  createCookieSessionStorage,
} from '@shopify/remix-oxygen';

/**
 * Export a fetch handler in module format.
 */
export default {
  async fetch(request, env, executionContext) {
    try {
      /**
       * Open a cache instance in the worker and a custom session instance.
       */
      if (!env?.SESSION_SECRET) {
        throw new Error('SESSION_SECRET environment variable is not set');
      }

      const waitUntil = (p) => executionContext.waitUntil(p);
      const [cache, session] = await Promise.all([
        caches.open('hydrogen'),
        HydrogenSession.init(request, [env.SESSION_SECRET]),
      ]);

      /**
       * Create Hydrogen's Storefront client.
       */
      const {storefront} = createStorefrontClient({
        cache,
        waitUntil,
        i18n: {language: 'EN', country: 'US'},
        publicStorefrontToken: env.PUBLIC_STOREFRONT_API_TOKEN,
        privateStorefrontToken: env.PRIVATE_STOREFRONT_API_TOKEN,
        storeDomain: `https://${env.PUBLIC_STORE_DOMAIN}`,
        storefrontApiVersion: env.PUBLIC_STOREFRONT_API_VERSION || '2023-01',
        storefrontId: env.PUBLIC_STOREFRONT_ID,
        storefrontHeaders: getStorefrontHeaders(request),
      });

      /**
       * Create a Remix request handler and pass
       * Hydrogen's Storefront client to the loader context.
       */
      const handleRequest = createRequestHandler({
        build: remixBuild,
        mode: process.env.NODE_ENV,
        getLoadContext: () => ({session, storefront, env}),
      });

      const response = await handleRequest(request);

      if (response.status === 404) {
        /**
         * Check for redirects only when there's a 404 from the app.
         * If the redirect doesn't exist, then `storefrontRedirect`
         * will pass through the 404 response.
         */
        return storefrontRedirect({request, response, storefront});
      }

      return response;
    } catch (error) {
      // eslint-disable-next-line no-console
      console.error(error);
      return new Response('An unexpected error occurred', {status: 500});
    }
  },
};

/**
 * This is a custom session implementation for your Hydrogen shop.
 * Feel free to customize it to your needs, add helper methods, or
 * swap out the cookie-based implementation with something else!
 */
class HydrogenSession {
  sessionStorage;
  session;
  constructor(sessionStorage, session) {
    this.sessionStorage = sessionStorage;
    this.session = session;
  }

  static async init(request, secrets) {
    const storage = createCookieSessionStorage({
      cookie: {
        name: 'session',
        httpOnly: true,
        path: '/',
        sameSite: 'lax',
        secrets,
      },
    });

    const session = await storage.getSession(request.headers.get('Cookie'));

    return new this(storage, session);
  }

  get(key) {
    return this.session.get(key);
  }

  destroy() {
    return this.sessionStorage.destroySession(this.session);
  }

  flash(key, value) {
    this.session.flash(key, value);
  }

  unset(key) {
    this.session.unset(key);
  }

  set(key, value) {
    this.session.set(key, value);
  }

  commit() {
    return this.sessionStorage.commitSession(this.session);
  }
}

My REMIX CONFIG file

/** @type {import('@remix-run/dev').AppConfig} */
module.exports = {
  appDirectory: 'app',
  ignoredRouteFiles: ['**/.*'],
  watchPaths: ['./public'],
  server: './server.js',
  /**
   * The following settings are required to deploy Hydrogen apps to Oxygen:
   */
  publicPath: (process.env.HYDROGEN_ASSET_BASE_URL ?? '/') + 'build/',
  assetsBuildDirectory: 'dist/client/build',
  serverBuildPath: 'dist/worker/index.js',
  serverMainFields: ['browser', 'module', 'main'],
  serverConditions: ['worker', process.env.NODE_ENV],
  serverDependenciesToBundle: 'all',
  serverModuleFormat: 'esm',
  serverPlatform: 'neutral',
  serverMinify: process.env.NODE_ENV === 'production',
};
my package .json 
{
  "name": "hello-world",
  "private": true,
  "sideEffects": false,
  "version": "0.0.0",
  "scripts": {
    "build": "npm run build:css && shopify hydrogen build",
    "build:css": "postcss app/styles/tailwind.css -o app/styles/tailwind-build.css ",
    "dev": "npm run build:css && concurrently -g -r npm:dev:css \"shopify hydrogen dev\"",
    "dev:css": "postcss app/styles/tailwind.css -o app/styles/tailwind-build.css -w"
  },
  "prettier": "@shopify/prettier-config",
  "dependencies": {
    "@headlessui/react": "^1.7.13",
    "@remix-run/react": "^1.14.0",
    "@shopify/cli": "3.45.0",
    "@shopify/cli-hydrogen": "^4.0.9",
    "@shopify/hydrogen": "^2023.1.6",
    "@shopify/remix-oxygen": "^1.0.4",
    "autoprefixer": "^10.4.14",
    "axios": "^0.24.0",
    "concurrently": "8.0.1",
    "graphql": "^16.6.0",
    "graphql-tag": "^2.12.6",
    "mpesa-api": "^3.0.2",
    "postcss": "^8.4.21",
    "postcss-cli": "^10.1.0",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "remix": "^1.15.0",
    "tailwindcss": "^3.3.1"
  },
  "devDependencies": {
    "@remix-run/dev": "^1.15.0",
    "@shopify/prettier-config": "^1.1.2",
    "eslint": "^8.37.0",
    "eslint-plugin-hydrogen": "0.12.3",
    "prettier": "^2.8.7"
  },
  "engines": {
    "node": ">=16.13"
  }
}
My vercel json file 

{
  "version": 2,
  "routes": [
    { "src": "/api/(.*)", "dest": "/api/$1" },
    { "src": "/", "dest": "/server.js" },
    { "src": "/(.*)", "dest": "/server.js" }
  ],
  "builds": [
    { "src": "server.js", "use": "@vercel/node" },
    {
      "src": "app/**",
      "use": "@vercel/static",
      "config": { "basePath": "dist/client" }
    },
    {
      "src": "./server",
      "use": "npm:@vercel/node",
      "config": { "basePath": "dist/worker", "entry": "./server" }
    },
    {
      "src": "./build.css",
      "use": "npm:build:css",
      "config": { "entry": "./build.css" }
    }
  ]
}
 my entry.server.jsxn file 

import {RemixServer} from '@remix-run/react';
import {renderToReadableStream} from 'react-dom/server';

export default async function handleRequest(
  request,
  responseStatusCode,
  responseHeaders,
  remixContext,
) {
  const body = await renderToReadableStream(
    <RemixServer context={remixContext} url={request.url} />,
  );

  responseHeaders.set('Content-Type', 'text/html');

  return new Response(body, {
    status: responseStatusCode,
    headers: responseHeaders,
  });
}
